src/modules/0.0.23/iasql_functions/iasql.ts

```diff
@@ -1039,3 +1038,878 @@ export async function continueUpgrade(
   });
   await install([...modsToInstall.values()], dbId, dbUser, false, true);
 }
+
+export async function commit(
+  dbId: string,
+  dryRun: boolean,
+  context: Context,
+  force = false,
+  ormOpt?: TypeormWrapper,
+) {
+  const t1 = Date.now();
+  logger.info(`Commiting to ${dbId}`);
+  await throwIfUpgrading(dbId, force);
+  const versionString = await TypeormWrapper.getVersionString(dbId);
+  const importedModules = await getImportedModules(dbId, versionString, force);
+  // We save the incoming orm from context if any and re-assign it before exiting
+  const parentOrm = context?.orm;
+  let orm: TypeormWrapper | null = null;
+  try {
+    orm = ormOpt ? ormOpt : await TypeormWrapper.createConn(dbId);
+    context.orm = orm;
+
+    const newStartCommit: IasqlAuditLog = await insertCommit(orm, dryRun ? 'preview_start' : 'start');
+    if (dryRun) context.previewStartCommit = newStartCommit;
+    else context.startCommit = newStartCommit;
+    const previousStartCommit = await getPreviousStartCommit(orm, newStartCommit.ts);
+
+    const changesToCommit = await getChangesToCommit(orm, newStartCommit, previousStartCommit);
+    const tablesWithChanges = [...new Set(changesToCommit.map(c => c.tableName))];
+
+    const installedModulesNames = (await orm.find(IasqlModule)).map((m: any) => m.name);
+    const installedModules: ModuleInterface[] = (Object.values(importedModules) as ModuleInterface[]).filter(
+      mod => installedModulesNames.includes(`${mod.name}@${mod.version}`),
+    );
+
+    const modulesWithChanges: ModuleInterface[] = getModulesWithChanges(
+      importedModules,
+      installedModules,
+      installedModulesNames,
+      tablesWithChanges,
+      versionString,
+    );
+
+    const modulesWithChangesSorted: ModuleInterface[] = sortModules(
+      modulesWithChanges,
+      installedModulesNames,
+    );
+    const installedModulesSorted: ModuleInterface[] = sortModules(installedModules, installedModulesNames);
+
+    const t2 = Date.now();
+    logger.info(`Setup took ${t2 - t1}ms`);
+
+    const crupdes: { toCreate: Crupde; toUpdate: Crupde; toReplace: Crupde; toDelete: Crupde } = {
+      toCreate: {},
+      toUpdate: {},
+      toReplace: {},
+      toDelete: {},
+    };
+
+    try {
+      if (modulesWithChanges.length) {
+        logger.info('Starting commit apply phase for modules with changes');
+        const applyRes = await commitApply(
+          dbId,
+          modulesWithChangesSorted,
+          context,
+          force,
+          crupdes,
+          dryRun,
+          changesToCommit,
+        );
+        if (dryRun) return applyRes;
+      }
+    } catch (e) {
+      logger.info('Something failed. Starting commit apply phase for all modules');
+      return await commitApply(dbId, installedModulesSorted, context, force, crupdes, dryRun);
```

Comment:
This behavior feels a little bit dangerous, as it could blow away changes in the cloud account that were not touched by the user in IaSQL.

What's the reasoning behind it?

Reply:
Well, I think we talked offline about this. This will help in cases where a dependency for a change we need to apply has been deleted or modified remotely outside of IaSQL. If we just take into account the changes, it will fail and on sync it will delete the user's change, so we do a full apply to restore resources in case of deletion in the cloud and see if the change we are trying to apply succeeds. This indeed could restore resources that might have been deleted from the cloud but we need to prioritize our DB as the SoT.

Reply:
I approved with this answer. I think the solution is going to be not fully satisfactory either way because a failure here is not going to 100% do what users want all of the time. I do agree the opposite behavior would be worse, though.